internal/status/httproutestatus.go

```diff
@@ -0,0 +1,150 @@
+
+type HTTPRouteUpdate struct {
+	FullName           types.NamespacedName
+	Conditions         []metav1.Condition
+	ExistingConditions []metav1.Condition
+	GatewayRef         types.NamespacedName
+	Generation         int64
+	TransitionTime     metav1.Time
+}
+
+// AddCondition returns a metav1.Condition for a given ConditionType.
+func (routeUpdate *HTTPRouteUpdate) AddCondition(cond gatewayapi_v1alpha1.RouteConditionType, status metav1.ConditionStatus, reason RouteReasonType, message string) metav1.Condition {
+	newDc := metav1.Condition{
+		Reason:             string(reason),
+		Status:             status,
+		Type:               string(cond),
+		Message:            message,
+		LastTransitionTime: metav1.NewTime(time.Now()),
+		ObservedGeneration: routeUpdate.Generation,
+	}
+	routeUpdate.Conditions = append(routeUpdate.Conditions, newDc)
+	return newDc
+}
+
+// HTTPRouteAccessor returns a HTTPRouteUpdate that allows a client to build up a list of
+// metav1.Conditions as well as a function to commit the change back to the cache when everything
+// is done. The commit function pattern is used so that the HTTPRouteUpdate does not need
+// to know anything the cache internals.
+func (c *Cache) HTTPRouteAccessor(route *gatewayapi_v1alpha1.HTTPRoute) (*HTTPRouteUpdate, func()) {
+	pu := &HTTPRouteUpdate{
+		FullName:           k8s.NamespacedNameOf(route),
+		Conditions:         []metav1.Condition{},
+		ExistingConditions: c.getGatewayConditions(route.Status.Gateways),
+		GatewayRef:         c.gatewayRef,
+		Generation:         route.Generation,
+	}
+
+	return pu, func() {
+		c.commitHTTPRoute(pu)
+	}
+}
+
+func (c *Cache) commitHTTPRoute(pu *HTTPRouteUpdate) {
+	if len(pu.Conditions) == 0 {
+		return
+	}
+	c.httpRouteUpdates[pu.FullName] = pu
+}
+
+func (routeUpdate *HTTPRouteUpdate) Mutate(obj interface{}) interface{} {
+	o, ok := obj.(*gatewayapi_v1alpha1.HTTPRoute)
+	if !ok {
+		panic(fmt.Sprintf("Unsupported %T object %s/%s in HTTPRouteUpdate status mutator",
+			obj, routeUpdate.FullName.Namespace, routeUpdate.FullName.Name,
+		))
+	}
+
+	httpRoute := o.DeepCopy()
+
+	var gatewayStatuses []gatewayapi_v1alpha1.RouteGatewayStatus
+	var conditionsToWrite []metav1.Condition
+
+	for _, cond := range routeUpdate.Conditions {
+
+		// Look through the newly calculated conditions, if they already exist
+		// on the object or if our obervation is stale, then leave them be,
+		// otherwise add them to the list to be written back to the API.
+		for _, existingCond := range routeUpdate.ExistingConditions {
+			if (cond.Status == existingCond.Status &&
+				cond.Reason == existingCond.Reason &&
+				cond.Message == existingCond.Message &&
+				cond.Type == existingCond.Type) ||
+				existingCond.ObservedGeneration > cond.ObservedGeneration {
+
+				cond.ObservedGeneration = existingCond.ObservedGeneration
+				cond.LastTransitionTime = existingCond.LastTransitionTime
+			}
+		}
+
+		conditionsToWrite = append(conditionsToWrite, cond)
```

Comment:
OK, I took a look at the `HTTPProxy` equivalent here, and I think there are a couple things:

- at the top of the loop, we should *always* set `cond.ObservedGeneration = routeUpdate.Generation` and `cond.LastTransitionTime = routeUpdate.LastTransitionTime`
- the inner loop should be looking for an existing Condition on the HTTPRoute that's got the same `Type`, but a newer observed generation. If we find one, that means `cond` is stale, so we should keep the existing one instead. If we don't find a newer existing one, then we should use `cond`.

So I think this should look more like (I wrote this in GitHub so it's definitely not 100% right):

```go
for _, cond := range routeUpdate.Conditions {
        // set the Condition's observed generation based on 
        // the generation of the HTTPRoute we looked at.
        cond.ObservedGeneration = routeUpdate.Generation
        cond.LastTransitionTime = routeUpdate.TransitionTime

		// is there a newer Condition on the HTTPRoute matching
		// this condition's type? If so, our observation is stale,
		// so don't write it, keep the newer one instead.
		var newerConditionExists bool
		for _, existingCond := routeUpdate.ExistingConditions {
			if existingCond.Type != cond.Type {
				continue
			}
			
			if existingCond.ObservedGeneration > cond.ObservedGeneration {
				conditionsToWrite = append(conditionsToWrite, existingCond)
				newerConditionExists = true
				break
			}
		}

		// if we didn't find a newer version of the Condition on the
		// HTTPRoute, then write the one we computed.
		if !newerConditionExists {
			conditionsToWrite = append(conditionsToWrite, cond)
		}
	}
```